content/blog/design-kgateway-for-scalability.md

```diff
@@ -0,0 +1,170 @@
+
+## Core Scalability Scenarios
+
+### 1. When a New Route Is Added
+
+In a production system with thousands of existing routes, adding a new one should be near-instant. Ideally, the route becomes effective within 1 second, without needing a periodic timer or restart. This ensures teams can test and deploy with confidence.
+
+{{< reuse-image src="blog/scalability-add-a-route.gif" width="750px" >}}
+
+### 2. When a Route Is Updated
+
+Updates to an existing route should not interrupt traffic. The old route should be replaced atomically, and the new route should take effect immediately. Any delay or downtime can lead to outages or security issues—especially in critical systems.
+
+{{< reuse-image src="blog/scalability-modify-a-route.gif" width="750px" >}}
+
+### 3. When a Route Is Removed
+
+Removing a route should instantly revoke access. Lingering routes, even for a few seconds, can pose security risks and violate compliance requirements.
+
+{{< reuse-image src="blog/scalability-delete-a-route.gif" width="750px" >}}
+
+### 4. When the Control Plane Is Restarted or Scaled
+
+The system should remain fully operational when the control plane restarts or scales out. Since the Envoy proxies are already configured, their behavior should not be affected. Once the control plane returns, it should not trigger unnecessary reconfigurations if there’s been no change.
+
+## Test These Scenarios Yourself
+
+1. Install kgateway.
+
+2. Create the `http` gateway proxy via the Gateway resource below:
+
+```yaml
+$ kubectl apply -f - <<EOF
+kind: Gateway
+apiVersion: gateway.networking.k8s.io/v1
+metadata:
+ name: http
+ namespace: kgateway-system
+spec:
+ gatewayClassName: kgateway
+ listeners:
+ - protocol: HTTP
+   port: 8080
+   name: http
+   allowedRoutes:
+     namespaces:
+       from: All
+EOF
+```
+
+3. **(Optional)** Install the [OpenTelemetry Collector](https://kgateway.dev/docs/observability/#otel) and [kube-prometheus-stack](https://kgateway.dev/docs/observability/#grafana) to observe CPU/memory and Envoy metrics. Make sure your cluster has adequate resources before doing this.
+
+4. Clone the [kgateway repo](https://github.com/kgateway-dev/kgateway). From the cloned directory, load 10,000 routes and backends (starting from 0 to 9999). Send a few random requests to confirm routes are effective immediately.
+
+```bash
+wget https://raw.githubusercontent.com/linsun/gateway-tests/refs/heads/main/scale/routes.yaml
+hack/utils/applier/go run main.go apply -f routes.yaml --iterations 10000
+kubectl port-forward deployment/http -n kgateway-system 8080:8080 &
+curl http://localhost:8080/foo/9
+curl http://localhost:8080/foo/99
+curl http://localhost:8080/foo/9999
+```
+
+5. **Add the 10,000th route and backend**, and check if the request to the newly added route is effective immediately.
```

Comment:
```suggestion
5. Add the 10,001st route and backend, and check if the request to the newly added route is effective immediately.
```

if they're zero based this will be 10,001

Reply:
i would be 10000 :) because the first cmd loads 0 to 9999.